File: internal/protocol/frame.go
// Package protocol defines the CloudBridge wire protocol for multiplexed
// stream transport over WebRTC DataChannel, QUIC, or Relay connections.
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

// Stream identifiers for well-known channels.
const (
	StreamControl StreamID = 0x0000
	StreamSSH     StreamID = 0x0001
	StreamShell   StreamID = 0x0002
	StreamRDP     StreamID = 0x0003
	StreamDocker  StreamID = 0x0004
	StreamVNC     StreamID = 0x0005
)

// Frame types carried in the Type field.
const (
	FrameData          FrameType = 0x01
	FrameOpenStream     FrameType = 0x02
	FrameCloseStream   FrameType = 0x03
	FrameWindowUpdate  FrameType = 0x04
	FramePing          FrameType = 0x05
	FramePong          FrameType = 0x06
)

// Header size: StreamID(2) + Type(1) + Length(4) = 7 bytes.
const FrameHeaderSize = 7

// MaxPayloadSize is the maximum payload length in a single frame.
const MaxPayloadSize = 64 * 1024 // 64 KiB

// StreamID identifies a multiplexed stream within a tunnel connection.
type StreamID uint16

// FrameType identifies the kind of frame.
type FrameType uint8

// Frame is the unit of transmission on the wire.
//
// Wire format:
//
//	+----------+--------+-----------+----------+
//	| StreamID | Type   | Length    | Payload  |
//	| 2 bytes  | 1 byte | 4 bytes   | N bytes  |
//	+----------+--------+-----------+----------+
type Frame struct {
	StreamID StreamID
	Type     FrameType
	Payload  []byte
}

// MarshalBinary encodes the frame into its wire format.
func (f *Frame) MarshalBinary() ([]byte, error) {
	if len(f.Payload) > MaxPayloadSize {
		return nil, fmt.Errorf("payload too large: %d > %d", len(f.Payload), MaxPayloadSize)
	}

	buf := make([]byte, FrameHeaderSize+len(f.Payload))
	binary.BigEndian.PutUint16(buf[0:2], uint16(f.StreamID))
	buf[2] = byte(f.Type)
	binary.BigEndian.PutUint32(buf[3:7], uint32(len(f.Payload)))
	copy(buf[7:], f.Payload)

	return buf, nil
}

// UnmarshalBinary decodes a frame from its wire format.
func (f *Frame) UnmarshalBinary(data []byte) error {
	if len(data) < FrameHeaderSize {
		return errors.New("frame too short")
	}

	f.StreamID = StreamID(binary.BigEndian.Uint16(data[0:2]))
	f.Type = FrameType(data[2])
	length := binary.BigEndian.Uint32(data[3:7])

	if len(data) < FrameHeaderSize+int(length) {
		return fmt.Errorf("frame truncated: expected %d bytes payload, got %d", length, len(data)-FrameHeaderSize)
	}
	if length > MaxPayloadSize {
		return fmt.Errorf("payload too large: %d > %d", length, MaxPayloadSize)
	}

	f.Payload = make([]byte, length)
	copy(f.Payload, data[FrameHeaderSize:FrameHeaderSize+length])

	return nil
}

// WriteTo writes the frame to an io.Writer in wire format.
// Returns the number of bytes written.
func (f *Frame) WriteTo(w io.Writer) (int64, error) {
	data, err := f.MarshalBinary()
	if err != nil {
		return 0, err
	}
	n, err := w.Write(data)
	return int64(n), err
}

// ReadFrom reads a frame from an io.Reader.
// It reads the header first, then the payload.
func ReadFrom(r io.Reader) (*Frame, error) {
	header := make([]byte, FrameHeaderSize)
	if _, err := io.ReadFull(r, header); err != nil {
		return nil, fmt.Errorf("reading frame header: %w", err)
	}

	streamID := StreamID(binary.BigEndian.Uint16(header[0:2]))
	frameType := FrameType(header[2])
	length := binary.BigEndian.Uint32(header[3:7])

	if length > MaxPayloadSize {
		return nil, fmt.Errorf("payload too large: %d > %d", length, MaxPayloadSize)
	}

	payload := make([]byte, length)
	if length > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return nil, fmt.Errorf("reading frame payload: %w", err)
		}
	}

	return &Frame{
		StreamID: streamID,
		Type:     frameType,
		Payload:  payload,
	}, nil
}

// String returns a human-readable description of the frame type.
func (t FrameType) String() string {
	switch t {
	case FrameData:
		return "DATA"
	case FrameOpenStream:
		return "OPEN_STREAM"
	case FrameCloseStream:
		return "CLOSE_STREAM"
	case FrameWindowUpdate:
		return "WINDOW_UPDATE"
	case FramePing:
		return "PING"
	case FramePong:
		return "PONG"
	default:
		return fmt.Sprintf("UNKNOWN(%d)", t)
	}
}

// String returns a human-readable description of the stream.
func (s StreamID) String() string {
	switch s {
	case StreamControl:
		return "control"
	case StreamSSH:
		return "ssh"
	case StreamShell:
		return "shell"
	case StreamRDP:
		return "rdp"
	case StreamDocker:
		return "docker"
	case StreamVNC:
		return "vnc"
	default:
		return fmt.Sprintf("stream-%d", s)
	}
}

File: app/lib/core/protocol/frame.dart
/// Frame protocol for CloudBridge multiplexed stream transport.
/// Aligned with Go's internal/protocol/frame.go
library;

import 'dart:typed_data';

/// Stream IDs for well-known channels
class StreamID {
  static const int control = 0x0000;
  static const int ssh = 0x0001;
  static const int shell = 0x0002;
  static const int rdp = 0x0003;
  static const int docker = 0x0004;
  static const int vnc = 0x0005;
}

/// Frame types
class FrameType {
  static const int data = 0x01;
  static const int openStream = 0x02;
  static const int closeStream = 0x03;
  static const int windowUpdate = 0x04;
  static const int ping = 0x05;
  static const int pong = 0x06;
}

/// Frame header size: StreamID(2) + Type(1) + Length(4) = 7 bytes
const int frameHeaderSize = 7;

/// Maximum payload size: 64 KiB
const int maxPayloadSize = 64 * 1024;

/// A frame in the CloudBridge wire protocol.
///
/// Wire format:
/// +----------+--------+-----------+----------+
/// | StreamID | Type   | Length    | Payload  |
/// | 2 bytes  | 1 byte | 4 bytes   | N bytes  |
/// +----------+--------+-----------+----------+
class Frame {
  final int streamId;
  final int type;
  final Uint8List payload;

  Frame({
    required this.streamId,
    required this.type,
    required this.payload,
  });

  /// Encode the frame to its wire format.
  Uint8List encode() {
    if (payload.length > maxPayloadSize) {
      throw ArgumentError(
          'Payload too large: ${payload.length} > $maxPayloadSize');
    }

    final buf = Uint8List(frameHeaderSize + payload.length);
    final view = ByteData.view(buf.buffer);

    // StreamID (2 bytes, big-endian)
    view.setUint16(0, streamId);
    // Type (1 byte)
    buf[2] = type;
    // Length (4 bytes, big-endian)
    view.setUint32(3, payload.length);

    // Payload
    buf.setRange(frameHeaderSize, frameHeaderSize + payload.length, payload);

    return buf;
  }

  /// Decode a frame from its wire format.
  static Frame decode(Uint8List data) {
    if (data.length < frameHeaderSize) {
      throw FormatException('Frame too short: ${data.length} < $frameHeaderSize');
    }

    final view = ByteData.view(data.buffer, data.offsetInBytes, data.length);
    final streamId = view.getUint16(0);
    final type = data[2];
    final length = view.getUint32(3);

    if (data.length < frameHeaderSize + length) {
      throw FormatException(
          'Frame truncated: expected $length bytes payload, got ${data.length - frameHeaderSize}');
    }
    if (length > maxPayloadSize) {
      throw FormatException('Payload too large: $length > $maxPayloadSize');
    }

    final payload = Uint8List.sublistView(
        data, frameHeaderSize, frameHeaderSize + length);

    return Frame(streamId: streamId, type: type, payload: payload);
  }

  @override
  String toString() => 'Frame(streamId=$streamId, type=$type, payloadLen=${payload.length})';
}

/// Read a frame from a byte buffer, returning the frame and remaining bytes.
/// Returns null if the buffer doesn't contain a complete frame.
FrameReadResult? tryReadFrame(Uint8List buffer) {
  if (buffer.length < frameHeaderSize) {
    return null; // Not enough data for header
  }

  final view = ByteData.view(buffer.buffer, buffer.offsetInBytes, buffer.length);
  final length = view.getUint32(3);

  if (buffer.length < frameHeaderSize + length) {
    return null; // Not enough data for payload
  }

  final frameData = Uint8List.sublistView(buffer, 0, frameHeaderSize + length);
  final frame = Frame.decode(frameData);
  final remaining = buffer.length - frameHeaderSize - length;

  return FrameReadResult(
    frame: frame,
    consumedBytes: frameHeaderSize + length,
    remainingBytes: remaining,
  );
}

class FrameReadResult {
  final Frame frame;
  final int consumedBytes;
  final int remainingBytes;

  FrameReadResult({
    required this.frame,
    required this.consumedBytes,
    required this.remainingBytes,
  });
}